### PR TITLE
Feature/self supervised

### PR DESCRIPTION
**Pull request by Actis92**
Add functionality for self supervised learning, in particular:
- Add new kind of task: ssl
- Add 2 type of ssl: Denoising and Contrastive
- Add 2 type of augmentations: cutmix, mixup
- Refactoring models, in order to have a clear distinction between backbone and head
- Add ssl_head that is used during in order to compute the output of the ssl task
- Remove some redundant code
- Add test
- Add comments

### CODE REVIEW

File: pytorch_tabular/tabular_model.py [hunk 1]
```diff
@@ -20,7 +20,6 @@
 from sklearn.base import TransformerMixin
 from torch import nn
 from tqdm.autonotebook import tqdm
-from pytorch_tabular import config
```
Comment (manujosephv) [inline pytorch_tabular/tabular_model.py:23]: do we need this import?

File: pytorch_tabular/tabular_datamodule.py [hunk 2]
```diff
@@ -103,6 +103,8 @@ def update_config(self) -> None:
             self.config.output_dim = len(self.config.target)
         elif self.config.task == "classification":
             self.config.output_dim = len(self.train[self.config.target[0]].unique())
+        elif self.config.task == "ssl":
+            self.config.output_dim = self.train.shape[1] - 1
```
Comment (manujosephv) [inline pytorch_tabular/tabular_datamodule.py:107]: should this line be :
```python
 self.config.output_dim = self.train.shape[1] - self.config.output_dim
```
?

File: pytorch_tabular/models/base_model.py [hunk 3]
```diff
@@ -61,13 +68,16 @@ def _build_network(self):
...
         if self.custom_loss is None:
-            try:
-                self.loss = getattr(nn, self.hparams.loss)()
-            except AttributeError as e:
-                logger.error(
-                    f"{self.hparams.loss} is not a valid loss defined in the torch.nn module"
-                )
-                raise e
+            if self.hparams.ssl_task == "Contrastive":
+                self.loss = loss_contrastive
```
Comment (manujosephv) [inline pytorch_tabular/models/base_model.py:72]: @Actis92  Do you think its a good idea to also issue a warning to the user that whatever loss they might have set will be ignored in an SSL task? Or may be include it in the docs that loss will be ignored in SSL. what do you think?

File: pytorch_tabular/models/base_model.py [hunk 4]
```diff
@@ -159,38 +169,69 @@ def calculate_metrics(self, y, y_hat, tag):
...
+
+    def compute_head(self, x: Tensor):
+        y_hat = self.head(x)
+        if (self.hparams.task == "regression") and (
+            self.hparams.target_range is not None
+        ):
+            for i in range(self.hparams.output_dim):
+                y_min, y_max = self.hparams.target_range[i]
+                y_hat[:, i] = y_min + nn.Sigmoid()(y_hat[:, i]) * (y_max - y_min)
+        if type(self.head) == nn.Identity:
```
Comment (manujosephv) [inline pytorch_tabular/models/base_model.py:189]: This condition for checking nn.Identity is to distinguish SSL workflow from regular? If yes, can we add a comment there?

Comment (manujosephv) [reply pytorch_tabular/models/base_model.py:189]: Or use the same hparams.task==ssl condition

File: pytorch_tabular/models/base_model.py [hunk 5]
```diff
@@ -159,38 +169,69 @@ def calculate_metrics(self, y, y_hat, tag):
...
+            y_hat = self(batch_augmented)["logits"]
+        else:
+            y = batch["target"]
+            y_hat = self(batch)["logits"]
+        return y, y_hat
+
     def training_step(self, batch, batch_idx):
-        y = batch["target"]
-        y_hat = self(batch)["logits"]
+        y, y_hat = self.shared_step(batch)
```
Comment (manujosephv) [inline pytorch_tabular/models/base_model.py:222]: Can we change the function name to something more intuitive and indicative of the action being done in the method? something like forward_pass or something like that?

File: pytorch_tabular/config/config.py [hunk 6]
```diff
@@ -513,7 +513,11 @@ def update_versions(self, name):
 class ModelConfig:
     """Base Model configuration
     Args:
-        task (str): Specify whether the problem is regression of classification.Choices are: regression classification
+        task (str): Specify whether the problem is regression of classification.Choices are: regression classification ssl
```
Comment (manujosephv) [inline pytorch_tabular/config/config.py:516]: comment can be changed to 
`
Specify whether the problem is regression, classification, or self supervised learning. Choices are: regression classification ssl
`

File: pytorch_tabular/config/config.py [hunk 7]
```diff
@@ -542,13 +546,29 @@ class ModelConfig:
...
         metadata={
             "help": "Specify whether the problem is regression of classification.",
-            "choices": ["regression", "classification"],
+            "choices": ["regression", "classification", "ssl"],
+        }
+    )
+    ssl_task: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": "Specify whether the problem is regression of classification.",
```
Comment (manujosephv) [inline pytorch_tabular/config/config.py:555]: metadata : help needs to be updated

File: pytorch_tabular/config/config.py [hunk 8]
```diff
@@ -542,13 +546,29 @@ class ModelConfig:
...
+            "choices": ["Denoising", "Contrastive", None],
+        }
+    )
+    aug_task: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": "Specify whether the problem is regression of classification.",
```
Comment (manujosephv) [inline pytorch_tabular/config/config.py:562]: metadata : help needs to be updated

File: pytorch_tabular/config/config.py [hunk 9]
```diff
@@ -598,38 +624,23 @@ def __post_init__(self):
...
                 else self.metrics_params
             )
+        elif self.task == "ssl":
+            assert self.ssl_task, "if task is ssl, ssl_task cannot be None"
+            assert self.aug_task, "if task is ssl, aug_task cannot be None"
+            if self.ssl_task == "Contrastive":
+                self.loss = "ContrastiveLoss" if self.loss is None else self.loss
+            else:
+                self.loss = "MSELoss" if self.loss is None else self.loss
+            self.metrics = (
```
Comment (manujosephv) [inline pytorch_tabular/config/config.py:634]: What happens with metrics for SSL? Should we restrict it to regression metrics?

File: pytorch_tabular/models/mixture_density/mdn.py [hunk 10]
```diff
@@ -134,19 +136,14 @@ def generate_point_predictions(self, pi, sigma, mu, n_samples=None):
 
 class BaseMDN(BaseModel, metaclass=ABCMeta):
     def __init__(self, config: DictConfig, **kwargs):
-        assert config.task == "regression", "MDN is only implemented for Regression"
-        assert config.output_dim == 1, "MDN is not implemented for multi-targets"
+        assert config.task in ["regression", "ssl"], "MDN is only implemented for Regression"
```
Comment (manujosephv) [inline pytorch_tabular/models/mixture_density/mdn.py:139]: I think MDN is not relevant for SSL task. Only thing that is different in MDN is the specific MDN Head which estimates probabilistic parameters..

**Review by manujosephv — COMMENTED** [review]
All in all, great work on the PR. And I really liked the refactoring you did with the head and the backbone...

Have added a few comments. If you can take a look at them it will be great.

File: pytorch_tabular/tabular_model.py (hunk 1 above)
Comment (Actis92) [reply pytorch_tabular/tabular_model.py:23]: According to my IDE this import is never used, because below there is 
`from pytorch_tabular.config import (
    DataConfig,
    ExperimentConfig,
    ExperimentRunManager,
    ModelConfig,
    OptimizerConfig,
    TrainerConfig,
)`
and in rest of the code the variable config is override by read some config file or receiving the config as argument of a function

File: pytorch_tabular/tabular_datamodule.py (hunk 2 above)
Comment (Actis92) [reply pytorch_tabular/tabular_datamodule.py]: We cannot use 
`self.config.output_dim = self.train.shape[1] - self.config.output_dim`
because we don't know if the model will be used for classification, regression, multi-target after the self supervised task, then self.config.output_dim is not defined. I have removed the number 1 and now is 
`self.config.output_dim = self.train.shape[1]`
this number is only used in order to set the output of the ssl task, so in this case it's the dimension of the MLP that can be any number, but usually it's equal to the original dimension of the input

File: pytorch_tabular/config/config.py (hunk 6 above)
Comment (Actis92) [reply pytorch_tabular/config/config.py]: Fixed

File: pytorch_tabular/config/config.py (hunk 7 above)
Comment (Actis92) [reply pytorch_tabular/config/config.py]: Fixed

File: pytorch_tabular/config/config.py (hunk 8 above)
Comment (Actis92) [reply pytorch_tabular/config/config.py]: Fixed

File: pytorch_tabular/models/mixture_density/mdn.py (hunk 10 above)
Comment (Actis92) [reply pytorch_tabular/models/mixture_density/mdn.py]: Fixed, and I have done the same change also for Tabnet, because in this moment it's not useful use the ssl task since the head of the tabnet is doing almost nothing

File: pytorch_tabular/models/base_model.py (hunk 4 above)
Comment (Actis92) [reply pytorch_tabular/models/base_model.py:191]: I have add a comment, the Identity is used for Tabnet, this means that we cannot divide the model in backbone and head, so we cannot extract backbone features

File: pytorch_tabular/models/base_model.py (hunk 5 above)
Comment (Actis92) [reply pytorch_tabular/models/base_model.py]: Fixed

File: pytorch_tabular/config/config.py (hunk 9 above)
Comment (Actis92) [reply pytorch_tabular/config/config.py:639]: In case of SSL I think that we can use the regression metrics, since it's all about a reconstruction loss

File: pytorch_tabular/tabular_model.py (hunk 1 above)
Comment (manujosephv) [reply pytorch_tabular/tabular_model.py:23]: cool.. then we can remove this import

File: pytorch_tabular/tabular_datamodule.py (hunk 2 above)
Comment (manujosephv) [reply pytorch_tabular/tabular_datamodule.py]: makes sense... anyways we will only be using the continuous and categorical features defined by the user. So maybe we can use those to get the required output dim?

File: pytorch_tabular/config/config.py (hunk 9 above)
Comment (manujosephv) [reply pytorch_tabular/config/config.py:639]: This ties back to the discussion we have on LinkedIn.. So maybe we can take this separately..

File: pytorch_tabular/tabular_datamodule.py (hunk 2 above)
Comment (Actis92) [reply pytorch_tabular/tabular_datamodule.py]: yes it make sense, now I have updated the code setting the output_dim as len(self.config.categorical_cols) + len(self.config.continuous_cols)

Comment (manujosephv) [reply pytorch_tabular/tabular_datamodule.py]: resolved

File: pytorch_tabular/models/mixture_density/mdn.py (hunk 10 above)
Comment (manujosephv) [reply pytorch_tabular/models/mixture_density/mdn.py]: Can we also remove the SSL related code in the MDN models?

File: tests/test_autoint.py [hunk 11]
```diff
@@ -142,52 +142,84 @@ def test_classification(
...
+            "AveOccup",
+            "Latitude",
+            "Longitude",
+        ],
+    ],
+)
+@pytest.mark.parametrize("categorical_cols", [["HouseAgeBin"]])
+@pytest.mark.parametrize("continuous_feature_transform", [None])
+@pytest.mark.parametrize("normalize_continuous_features", [True])
+@pytest.mark.parametrize("target_range", [True, False])
```
Comment (manujosephv) [inline tests/test_autoint.py:163]: We can reduce the parameters a bit here... for eg. target_range will have no impact on the ssl task. Same feedback in all the other ssl test cases as well

File: tests/test_categorical_embedding.py [hunk 12]
```diff
@@ -205,55 +205,97 @@ def test_embedding_transformer(regression_data):
...
+            "Longitude",
+        ],
+        [],
+    ],
+)
+@pytest.mark.parametrize("categorical_cols", [["HouseAgeBin"], []])
+@pytest.mark.parametrize("continuous_feature_transform", [None, "yeo-johnson"])
+@pytest.mark.parametrize("normalize_continuous_features", [True, False])
+@pytest.mark.parametrize("target_range", [True, False])
+@pytest.mark.parametrize(
```
Comment (manujosephv) [inline tests/test_categorical_embedding.py:227]: target_transform, target_range, custom_metrics, cutom_loss, custom_optimizer, all can be set to default values

File: tests/test_mdn.py [hunk 13]
```diff
@@ -120,53 +120,69 @@ def test_classification(
...
+    ],
+)
+@pytest.mark.parametrize("categorical_cols", [["HouseAgeBin"]])
+@pytest.mark.parametrize("continuous_feature_transform", [None])
+@pytest.mark.parametrize("normalize_continuous_features", [True])
+@pytest.mark.parametrize("variant", [CategoryEmbeddingMDNConfig, NODEMDNConfig, AutoIntMDNConfig])
+@pytest.mark.parametrize("num_gaussian", [1, 2])
+@pytest.mark.parametrize("ssl_task", ["Denoising", "Contrastive"])
+@pytest.mark.parametrize("aug_task", ["cutmix", "mixup"])
+def test_ssl(
```
Comment (manujosephv) [inline tests/test_mdn.py:137]: No need for test_ssl for MDN

File: tests/test_tabnet.py [hunk 14]
```diff
@@ -120,77 +120,50 @@ def test_classification(
...
+    [
+        [f"feature_{i}" for i in range(54)],
+    ],
+)
+@pytest.mark.parametrize("categorical_cols", [["feature_0_cat"]])
+@pytest.mark.parametrize("continuous_feature_transform", [None])
+@pytest.mark.parametrize("normalize_continuous_features", [True])
+@pytest.mark.parametrize("ssl_task", ["Denoising", "Contrastive"])
+@pytest.mark.parametrize("aug_task", ["cutmix", "mixup"])
+def test_ssl(
```
Comment (manujosephv) [inline tests/test_tabnet.py:135]: I think ssl does not work for Tabnet as well, right? thenwe can take this test case off as well

**Review by manujosephv — CHANGES_REQUESTED** [review]
Have added a few more comments, mainly in test cases

File: tests/test_autoint.py (hunk 11 above)
Comment (Actis92) [reply tests/test_autoint.py]: I have removed some of the test cases that have no impact on ssl

File: tests/test_categorical_embedding.py (hunk 12 above)
Comment (Actis92) [reply tests/test_categorical_embedding.py:226]: Fixed

File: pytorch_tabular/models/mixture_density/mdn.py (hunk 10 above)
Comment (Actis92) [reply pytorch_tabular/models/mixture_density/mdn.py]: I have removed the related code about ssn

File: tests/test_mdn.py (hunk 13 above)
Comment (Actis92) [reply tests/test_mdn.py:136]: I have keep only one test in order to check the raise of the exception in case of ssl task

File: tests/test_tabnet.py (hunk 14 above)
Comment (Actis92) [reply tests/test_tabnet.py:135]: I have keep only one test in order to check the raise of the exception in case of ssl task

**Review by manujosephv — APPROVED** [review]
